Missing defines on systems with kernel < 4.X

## Description
We are running into trouble with centos 7 based builds of snmalloc. ( [Dockerimage](https://hub.docker.com/layers/centos/library/centos/centos7.9.2009/images/sha256-8faead07bd1d5fdd17414a1759eae004da5daa9575a846f4730f44dec0f47843) )

We do have to support such an old system for the time being (don't ask), you might not want to support it, which i totally understand.
The system is based on gcc 9, kernel 3.10.0 and libraries, which is all horribly outdated.

Our build system gives the following error:

```
...
 /cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.3.1/snmalloc/src/snmalloc/override/../global/../backend/../backend_helpers/../mem/../ds/../pal/pal_linux.h: In static member function 'static void snmalloc::PALLinux::notify_not_using(void*, size_t)':
  /cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.3.1/snmalloc/src/snmalloc/override/../global/../backend/../backend_helpers/../mem/../ds/../pal/pal_linux.h:111:24: error: 'MADV_FREE' was not declared in this scope; did you mean 'MADV_REMOVE'?
    111 |       madvise(p, size, MADV_FREE);
        |                        ^~~~~~~~~
        |                        MADV_REMOVE
  /cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.3.1/snmalloc/src/snmalloc/override/../global/../backend/../backend_helpers/../mem/../ds/../pal/pal_linux.h: In static member function 'static uint64_t snmalloc::PALLinux::get_entropy64()':
  /cargo/registry/src/github.com-1ecc6299db9ec823/snmalloc-sys-0.3.1/snmalloc/src/snmalloc/override/../global/../backend/../backend_helpers/../mem/../ds/../pal/pal_linux.h:180:57: error: 'GRND_NONBLOCK' was not declared in this scope; did you mean 'O_NONBLOCK'?
    180 |           ret = syscall(SYS_getrandom, current, length, GRND_NONBLOCK);
        |                                                         ^~~~~~~~~~~~~
        |                                                         O_NONBLOCK
  gmake[3]: *** [CMakeFiles/snmallocshim-rust.dir/src/snmalloc/override/rust.cc.o] Error 1
```

## Possible Solutions

On those old systems we do have `GRND_NONBLOCK` etc. defined in `/usr/include/linux/random.h`, i do think this include is missing (maybe guarded by a check if `GRND_NONBLOCK` is defined).

`MADV_FREE` is only available since linux 4.5. Should we fallback to `MADV_REMOVE` for systems where `MADV_FREE` is not defined?

I will happily provide those fixes if you deem them good enough or maybe you have another idea?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,7 +8,7 @@ endif()
 
 include(CheckCXXCompilerFlag)
 include(CheckCXXSourceCompiles)
-include(CheckIncludeFile)
+include(CheckIncludeFileCXX)
 include(CMakeDependentOption)
 
 option(SNMALLOC_HEADER_ONLY_LIBRARY "Use snmalloc has a header-only library" OFF)
@@ -112,7 +112,7 @@ int main() {
 # but some might provide the necessary flags via linux/random.h
 # the __has_include macro isn't working properly on all platforms for that header
 # this is why we check its existence here
-CHECK_INCLUDE_FILE(linux/random.h SNMALLOC_HAS_LINUX_RANDOM_H)
+CHECK_INCLUDE_FILE_CXX(linux/random.h SNMALLOC_HAS_LINUX_RANDOM_H)
 
 # Provide as function so other projects can reuse
 # FIXME: This modifies some variables that may or may not be the ones that

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,6 +8,7 @@ endif()
 
 include(CheckCXXCompilerFlag)
 include(CheckCXXSourceCompiles)
+include(CheckIncludeFile)
 include(CMakeDependentOption)
 
 option(SNMALLOC_HEADER_ONLY_LIBRARY "Use snmalloc has a header-only library" OFF)
@@ -105,6 +106,14 @@ int main() {
   return res;
 }
 " SNMALLOC_PLATFORM_HAS_GETENTROPY)
+
+# check if linux/random.h is available
+# older libcs might not have sys/random.h
+# but some might provide the necessary flags via linux/random.h
+# the __has_include macro isn't working properly on all platforms for that header
+# this is why we check its existence here
+CHECK_INCLUDE_FILE(linux/random.h SNMALLOC_HAS_LINUX_RANDOM_H)
+
 # Provide as function so other projects can reuse
 # FIXME: This modifies some variables that may or may not be the ones that
 # provide flags and so is broken by design.  It should be removed once Verona
@@ -203,6 +212,7 @@ add_as_define(SNMALLOC_QEMU_WORKAROUND)
 add_as_define(SNMALLOC_TRACING)
 add_as_define(SNMALLOC_CI_BUILD)
 add_as_define(SNMALLOC_PLATFORM_HAS_GETENTROPY)
+add_as_define(SNMALLOC_HAS_LINUX_RANDOM_H)
 if (SNMALLOC_NO_REALLOCARRAY)
   add_as_define(SNMALLOC_NO_REALLOCARRAY)
 endif()

--- a/src/snmalloc/pal/pal_linux.h
+++ b/src/snmalloc/pal/pal_linux.h
@@ -40,6 +40,19 @@ namespace snmalloc
      */
     static constexpr int default_mmap_flags = MAP_NORESERVE;
 
+    /**
+     * MADV_FREE is only available since Linux 4.5.
+     *
+     * Fallback to MADV_DONTNEED on older kernels
+     */
+    static constexpr int madvise_free_flags =
+#  ifdef SNMALLOC_HAS_LINUX_RANDOM_H
+      MADV_FREE
+#  else
+      MADV_DONTNEED
+#  endif
+      ;
+
     static void* reserve(size_t size) noexcept
     {
       void* p = PALPOSIX<PALLinux>::reserve(size);
@@ -113,15 +126,7 @@ namespace snmalloc
         memset(p, 0x5a, size);
 
       madvise(p, size, MADV_DONTDUMP);
-      madvise(
-        p,
-        size,
-#  if defined(MADV_FREE)
-        MADV_FREE
-#  else
-        MADV_DONTNEED
-#  endif
-      );
+      madvise(p, size, madvise_free_flags);
 
       if constexpr (PalEnforceAccess)
       {

--- a/src/snmalloc/pal/pal_linux.h
+++ b/src/snmalloc/pal/pal_linux.h
@@ -8,7 +8,9 @@
 #  include <sys/mman.h>
 #  include <sys/prctl.h>
 #  include <syscall.h>
-#  if !defined(GRND_NONBLOCK) && __has_include(<linux/random.h>)
+// __has_include does not reliably determine if we actually have linux/random.h
+// available
+#  if defined(SNMALLOC_HAS_LINUX_RANDOM_H)
 #    include <linux/random.h>
 #  endif
 

--- a/src/snmalloc/pal/pal_linux.h
+++ b/src/snmalloc/pal/pal_linux.h
@@ -8,6 +8,9 @@
 #  include <sys/mman.h>
 #  include <sys/prctl.h>
 #  include <syscall.h>
+#  if !defined(GRND_NONBLOCK) && __has_include(<linux/random.h>)
+#    include <linux/random.h>
+#  endif
 
 extern "C" int puts(const char* str);
 
@@ -108,7 +111,15 @@ namespace snmalloc
         memset(p, 0x5a, size);
 
       madvise(p, size, MADV_DONTDUMP);
-      madvise(p, size, MADV_FREE);
+      madvise(
+        p,
+        size,
+#  if defined(MADV_FREE)
+        MADV_FREE
+#  else
+        MADV_DONTNEED
+#  endif
+      );
 
       if constexpr (PalEnforceAccess)
       {

